### PR TITLE
Tail recursion warning and attribute, take 2

### PR DESCRIPTION
**Pull request by dawedawe**
The last time this was [tried](https://github.com/dotnet/fsharp/pull/1976) it didn't really work out.
But as there is an [RFC](https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1011-warn-on-recursive-without-tail-call.md) it should be worth to work on.
So, to test the waters, here's a very early draft based on the work of @AviAvni 
It's doing pretty much the same but ~~uses a field in cenv instead of a function parameter to pass around.~~ collects the attributed vals in a different way.

There's still a lot to do and I'll need help from more experienced people. Given the amount of support for the proposal last time, I hope we can make more progress this time.

### CODE REVIEW

**Comment by Happypig375** [comment]
per https://github.com/fsharp/fslang-design/discussions/82#discussioncomment-6007931, I don't think an attribute is the way to go.

**Comment by dsyme** [comment]
Please see this discussion comment: https://github.com/fsharp/fslang-design/discussions/82#discussioncomment-6011662

**Comment by vzarytovskii** [comment]
This should also be under a language version flag.

**Comment by dawedawe** [comment]
> This should also be under a language version flag.

Yes, absolutely. Thanks.

File: src/Compiler/Checking/PostInferenceChecks.fs [hunk 1]
```diff
@@ -87,6 +87,8 @@ type env =
 
       /// Values in this recursive scope that have been marked [<TailCall>]
       mutable mustTailCall: Zset<Val>
+      
+      mutable mustTailCallRanges: Map<string, Range>
```
[the remark this answers is not in the document]
Comment (dawedawe) [reply src/Compiler/Checking/PostInferenceChecks.fs]: Nice, thanks!

**Comment by dawedawe** [comment]
For those following along, I thing this PR is superseded by https://github.com/dotnet/fsharp/pull/15503
Please go there for the lastest work.

**Pull request closed without merging**